src/main/java/com/minecolonies/entity/ai/citizen/lumberjack/Tree.java

```diff
@@ -162,59 +164,88 @@ private void checkTree(@NotNull World world, @NotNull BlockPos topLog)
     }
 
     /**
-     * Get's the top log of the tree
+     * For use in PathJobFindTree.
      *
-     * @param world The entity world
-     * @param pos   The coordinates
-     * @return the top log position
+     * @param world the world.
+     * @param pos   The coordinates.
+     * @return true if the log is part of a tree.
      */
-    private static BlockPos getTopLog(@NotNull IBlockAccess world, BlockPos pos)
+    public static boolean checkTree(@NotNull IBlockAccess world, BlockPos pos)
     {
-        BlockPos topPos = pos;
-        while (world.getBlockState(topPos.up()).getBlock().isWood(world, topPos.down()))
+        //Is the first block a log?
+        if (!world.getBlockState(pos).getBlock().isWood(world, pos))
         {
-            topPos = topPos.up();
+            return false;
         }
-        return topPos;
+
+        final Tuple<BlockPos, BlockPos> baseAndTOp = getBottomAndTopLog(world, pos, new LinkedList<>(), null, null);
+
+        //Get base log, should already be base log.
+        final BlockPos basePos = baseAndTOp.getFirst();
+
+        //Make sure tree is on solid ground and tree is not build above cobblestone.
+        return world.getBlockState(basePos.down()).getMaterial().isSolid()
+                 && world.getBlockState(basePos.down()).getBlock() != Blocks.COBBLESTONE
+                 && hasEnoughLeaves(world, baseAndTOp.getSecond());
     }
 
     /**
-     * For use in PathJobFindTree
+     * Adds a log and searches for further logs(Breadth first search).
      *
-     * @param world the world
-     * @param pos   The coordinates
-     * @return true if the log is part of a tree
+     * @param world The world the log is in.
+     * @param log   the log to add.
      */
-    public static boolean checkTree(@NotNull IBlockAccess world, BlockPos pos)
+    @NotNull
+    private static Tuple<BlockPos, BlockPos> getBottomAndTopLog(@NotNull IBlockAccess world, @NotNull BlockPos log, @NotNull LinkedList<BlockPos> woodenBlocks, BlockPos bottomLog,
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "getBottomAndTopLog" is 13 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)